// @nsa/nsa-cassandra/src/common/constants/message.js
/**
 * Created by magesh on 1/17/17.
 */
module.exports = Object.freeze({

    CONTACT_ADMIN: "Contact Admin",

    /**
     * Fee Type
     */
    FETCH_ALL_FEE_TYPES: 'fetch all the fee types',
    FETCH_FEE_TYPE: 'fetch the fee type',
    FEE_TYPE_SUCCESS : "Fee Type successfully saved",
    FEETYPE_ERROR : "Could not save Fee Type!!!",

    /**
     * Fee Scholarship
     */
    FETCH_ALL_FEE_SCHOLARSHIP: 'fetch all the fee scholarships',
    FETCH_FEE_SCHOLARSHIP: 'fetch the fee scholarship',
    FEE_SCHOLARSHIP_SUCCESS : "Fee Scholarship successfully saved",
    FEE_SCHOALRSHIP_ERROR : "Could not save Fee Scholarship!!!",

    /**
     * Fee Structure
     */
    FETCH_ALL_FEE_STRUCTURE: 'fetch all the fee structures',
    FETCH_FEE_STRUCTURE: 'fetch the fee structure',
    FEE_STRUCTURE_SUCCESS : "Fee Structure successfully saved",
    FEE_STRUCTURE_ERROR : "Could not save Fee Structure!!!",

    /**
     * Fee Assignment
     */
    FETCH_ALL_FEE_ASSIGNMENT: 'fetch all the fees',
    FETCH_FEE_ASSIGNMENT: 'fetch the fee',
    FEE_ASSIGNMENT_SUCCESS : "Fee successfully saved",
    FEE_ASSIGNMENT_ERROR : "Could not save Fee!!!",

    /**
     * Fee Assignment Details
     */
    FETCH_ALL_FEE_ASSIGNMENT_DETAILS: 'fetch all the fee assignment details',
    FETCH_FEE_ASSIGNMENT_DETAIL: 'fetch the fee assignment details',
    FEE_ASSIGNMENT_DETAIL_SUCCESS : "Fee Assignment Details successfully saved",
    FEE_ASSIGNMENT_DETAIL_ERROR : "Could not save Fee Assignment Details!!!",

    FAILURE : 'The request was invalid, please contact administrator',

    USER_SAVE_SUCCESS: "Successfully Saved",
    USER_SAVE_FAILURE: "Could not save user!!!",
    USER_GET_FAILURE: "cant get user details",
    USER_UPDATE_FAILURE: "Update failed",
    USER_UPDATE_SUCCESS: "Updated Successfully",

    TEMPLATE_SAVE_SUCCESS: "Template successfully saved",
    TEMPLATE_DELETE_SUCCESS: "Template deleted successsfully",
    TEMPLATE_UPDATE_SUCCESS: "Template updated successsfully"
});
