// @nsa/nsa-assessment/index.js
module.exports =  {
    employee : require('./models/employee.js'),
    module : require('./models/module.js'),
    classDetails: require('./models/ClassDetails.js'),
    subTopics: require('./models/SubTopics.js'),
    section: require('./models/section.js'),
    subjectTitle: require('./models/subjectTitle.js'),
    subjectTitleTerm: require('./models/subjectTitleTerm.js'),
    subjectTermTopics: require('./models/subjectTermTopics.js'),
    schoolSubjectTitle: require('./models/schoolSubjectTitle.js'),
    schedule: require('./models/schedule.js'),
    title: require('./models/title.js'),
    profile: require('./models/profile.js'),
    relatedStatus: require('./models/relatedStatus.js'),
    user: require('./models/user.js'),
    workflow: require('./models/workflow.js'),
    images: require('./models/images.js'),
    routing: require('./models/routing.js'),
    vCourse: require('./models/vCourse.js'),
    vCenter: require('./models/vCenter.js'),
    vCenterCourse: require('./models/vCenterCourse.js'),
    vSubject: require('./models/vSubject.js'),
    vSubjectTopics: require('./models/vSubjectTopics'),
    vTopic: require('./models/vTopic.js'),
    vBatch: require('./models/vBatch.js'),
    vBatchSchedule: require('./models/vBatchSchedule.js'),
    vBatchAssessmentSchedule: require('./models/vBatchAssessmentSchedule.js'),
    vAssignBatch: require('./models/vAssignBatch.js'),
    vFaculty: require('./models/vFaculty.js'),
    vFaculty: require('./models/vClass.js'),
    vAssessments: require('./models/vAssessments.js'),
    vAssessmentDetails: require('./models/vAssessmentDetails.js'),
    vStudent: require('./models/vStudent.js'),
    vStudentDetails: require('./models/vStudentDetails.js'),
    vStudentLeave: require('./models/vStudentLeave.js'),
    vFeeTypes: require('./models/vFeeTypes.js'),
    vRegistration : require('./models/vRegistration.js'),
    vFees: require('./models/vFees.js'),
    vFeeCard : require('./models/vFeeCard.js'),
    vTransactionReqDetails: require('./models/vTransactionReqDetails.js'),
    vTransactionResDetails: require('./models/vTransactionResDetails.js'),
    vFeeCardDetails: require('./models/vFeeCardDetails.js'),
    vStudentFeeDetails: require('./models/vStudentFeeDetails.js'),
    vStudentNotifications: require('./models/vStudentNotifications'),
    vEmployeeNotifications: require('./models/vEmployeeNotifications'),
    vOtherNotifications: require('./models/vOtherNotifications'),
    vTransacations: require('./models/vTransacations'),
    vQuestion: require('./models/vQuestion'),
    vSheet: require('./models/vSheet'),
    vExamConfig: require('./models/vExamConfig'),
    vExam: require('./models/vExam'),
    vCodeType:  require('./models/vCodeType'),
    vPracticeMetrics: require('./models/vPracticeMetrics'),
    vSubjectMetrics: require('./models/vSubjectMetrics'),
    vExamSchedule: require('./models/vExamSchedule'),
    vStudentExam: require('./models/vStudentExam'),
    vPracticeSchedule: require('./models/vPracticeSchedule'),
    vStudentPractice: require('./models/vStudentPractice'),
    vQuestionMetrics: require('./models/vQuestionMetrics'),
    vExamScheduleLog: require('./models/vExamScheduleLog'),
    vStudyContent: require('./models/vStudyContent'),
    testConfiguration: require('./models/testConfiguration'),
    questionUsage: require('./models/questionUsage'),
    '7pmSubjectMetrics':  require('./models/7pmSubjectMetrics'),
    '7pmStudentExam': require('./models/7pmStudentExam'),
    '7pmQuestionMetrics': require('./models/7pmQuestionMetrics')

};
